compiler/grainformat/comment_utils.re

```diff
@@ -0,0 +1,564 @@
+        (line: int, char: int, comments: list(Parsetree.comment)) =>
+  switch (comments) {
+  | [] => []
+  | [c, ...remaining_comments] =>
+    let c_loc: Location.t = Locations.get_comment_loc(c);
+    let (_, cmtline, cmtchar, _) =
+      Locations.get_raw_pos_info(c_loc.loc_start);
+
+    if (cmtline > line) {
+      []; // can stop early as there will be no more
+    } else if (cmtline == line && cmtchar >= char) {
+      [c, ...get_comments_on_line_end(line, char, List.tl(comments))];
+    } else {
+      get_comments_on_line_end(line, char, List.tl(comments));
+    };
+  };
+
+let rec get_comments_on_line_start = (line: int, char: int, comments) =>
+  switch (comments) {
+  | [] => []
+  | [c, ...remaining_comments] =>
+    let c_loc: Location.t = Locations.get_comment_loc(c);
+    let (_, cmtline, cmtchar, _) =
+      Locations.get_raw_pos_info(c_loc.loc_start);
+
+    if (cmtline > line) {
+      []; // can stop early as there will be no more
+    } else if (cmtline == line && cmtchar <= char) {
+      [c, ...get_comments_on_line_start(line, char, List.tl(comments))];
+    } else {
+      get_comments_on_line_start(line, char, List.tl(comments));
+    };
+  };
+
+let get_comments_to_end_of_line =
+    (~location: Location.t, comments: list(Parsetree.comment)) => {
+  let (_, stmt_end_line, stmt_end_char, _) =
+    Locations.get_raw_pos_info(location.loc_end);
+
+  get_comments_on_line_end(stmt_end_line, stmt_end_char, comments);
+};
+
+let comment_to_doc = (comment: Parsetree.comment) => {
+  let comment_string = Comments.get_comment_source(comment);
+  let newline =
+    switch (comment) {
+    | Line(_)
+    | Shebang(_) => Doc.hardLine
+    | Doc(_) => Doc.hardLine
+    | _ => Doc.nil
+    };
+  Doc.concat([Doc.text(String.trim(comment_string)), newline]);
+};
+
+let nobreak_comment_to_doc = (comment: Parsetree.comment) => {
+  let comment_string = Comments.get_comment_source(comment);
+
+  Doc.text(String.trim(comment_string));
+};
+
+let get_after_brace_comments =
+    (
+      loc: Location.t,
+      comments: list(Parsetree.comment),
+      first: option(Location.t),
+    ) => {
+  let (_, startline, startc, _) = Locations.get_raw_pos_info(loc.loc_start);
+
+  let cmts = get_comments_on_line(startline, comments);
+  switch (cmts) {
+  | [] => cmts
+  | [fst, ...rem] =>
+    switch (first) {
+    | None => cmts
+    | Some(leading) =>
+      let fstclog = Locations.get_comment_loc(fst);
+
+      let (_, itemstartline, itemstartc, _) =
+        Locations.get_raw_pos_info(leading.loc_start);
+
+      if (itemstartline > startline) {
+        cmts;
+      } else {
+        let (_, cmtstartline, cmtstartc, _) =
+          Locations.get_raw_pos_info(fstclog.loc_start);
+        if (cmtstartline >= itemstartline && cmtstartc > itemstartc) {
+          [];
+        } else {
+          cmts;
+        };
+      };
+    }
+  };
+};
+
+let rec comments_inner =
+        (
+          prev: option(Parsetree.comment),
```

Comment:
Why isn't this an optional, named param?

Reply:
Fixed.